data/transactions/logic/assembler.go

```diff
@@ -1153,19 +1147,21 @@ func asmAcctParams(ops *OpStream, spec *OpSpec, args []string) error {
 	return nil
 }
 
-func asmTxField(ops *OpStream, spec *OpSpec, args []string) error {
+func asmItxnField(ops *OpStream, spec *OpSpec, args []string) error {
 	if len(args) != 1 {
 		return ops.errorf("%s expects one argument", spec.Name)
 	}
-	fs, ok := TxnFieldSpecByName[args[0]]
+	fs, ok := txnFieldSpecByName[args[0]]
 	if !ok {
 		return ops.errorf("%s unknown field: %#v", spec.Name, args[0])
 	}
 	if fs.itxVersion == 0 {
-		return ops.errorf("%s %#v is not allowed.", spec.Name, args[0])
+		//nolint:errcheck // we continue to maintain typestack
+		ops.errorf("%s %#v is not allowed.", spec.Name, args[0])
 	}
 	if fs.itxVersion > ops.Version {
-		return ops.errorf("%s %#v available in version %d. Missed #pragma version?", spec.Name, args[0], fs.itxVersion)
+		//nolint:errcheck // we continue to maintain typestack
+		ops.errorf("%s %s field was introduced in TEAL v%d. Missed #pragma version?", spec.Name, args[0], fs.itxVersion)
```

Comment:
@jannotti A couple of beginner's questions here:

It seems the objective is to achieve a side-effect (`ops.Errors = append(ops.Errors, le)`) without returning an error to the caller.  Is that accurate?

If so, I am wondering:
* Can we instead provide a method that achieves the same effect without needing to suppress the linter?
  * I realize it's minor.  But, the pattern appears a number of times.
  * Options here include a that performs the side-effect with no return value.  Or a method that returns the error without the side-effect, and allows call-site to perform side effect.
* Secondarily, the highlighted error message repeats ~10 times.  So that we minimize the chance for having inconsistent error messaging, are you opposed to refactoring to remove the duplication?

It seems the patterns predate the PR.  _If_ we agree a change is beneficial and _if_ you prefer, I am on-board to defer to a subsequent PR. 


-----

More fundamentally:  What's the rationale for allowing these classes of errors to _continue_ assembly?

If easier to discuss live, happy to do so.

Reply:
You're going to be very happy with the PR that follows this one, which gets rid of most asmOpcode routines, so this will only appear once.

But I stand behind the design of ops.error(), mostly.  In an assembler/compiler tool, you mostly want to accumulate errors, rather than stop at the first one. So ops.errorf() is mostly about accumulation. But, it is very convenient to have it return the error, because then you can then use it as a return value in places where you don't want to continue.  In the body of assembling most opcodes, you mostly want to stop executing the body, and get back to the main loop, so you just return error created by ops.errorf().  But sometimes you want to continue, because you want to "maintain the typestack" (see comment after nolint:errcheck).  Some opcodes have a declared return type of StackAny, but once you know the field being accessed, you know the exact type being returned.  ops.returns() is a way for opcode assembly to say "Although my return type is generally Any, I now know my return type".  That happens at the end of this function, so it's better too proceed and get that info, so follow-on error messages can be more exact.

Now, the assembly opcode errors are always ignored, so I suppose I could make them return `void`, the use a void returning ops.errorf, and `return`.  Probably reasonable.  Maybe it's worth the extra line, and avoid confusing linters.

Reply:
@jannotti Thanks for the context.  Summarizing a live discussion below on thread questions.  Resolving.

> You're going to be very happy with the PR that follows this one

Sounds good.  ☕ 

> Now, the assembly opcode errors are always ignored,

* Thanks for pointing it out.  I hadn't realized until now.
* During our discussion, we agreed _for the next PR_:  Invoke `ops.returns(fs.ftype)` earlier to allow subsequent `ops.errorf` to safely return.  The proposal removes the linter error while maintaining the typestack.